src/mame/drivers/jungleyo.cpp

```diff
@@ -90,86 +112,179 @@ class jungleyo_state : public driver_device
-
-	if (data & 0xbfff)
-		LOGOUTPUTS("%s output_w: %04x\n", machine().describe_context(), data);
+	u16 code = m_bg_videoram[tile_index*2+1];
+	u16 color = m_bg_videoram[tile_index*2];
+	tileinfo.set(1, code, color & 0x1f, 0);
 }
 
-
-TILE_GET_INFO_MEMBER(jungleyo_state::get_bg_tile_info)
+TILE_GET_INFO_MEMBER(jungleyo_state::get_fg_tile_info)
 {
-	uint16_t code = m_bg_videoram[tile_index];
-	tileinfo.set(1, code, 0, 0);
+	u16 code = m_fg_videoram[tile_index*2+1];
+	u16 color = m_fg_videoram[tile_index*2];
+	tileinfo.set(2, code, color & 0x1f, 0);
 }
 
-TILE_GET_INFO_MEMBER(jungleyo_state::get_fg_tile_info)
+template <int layer_num> TILE_GET_INFO_MEMBER(jungleyo_state::get_reel_tile_info)
 {
-	uint16_t code = m_fg_videoram[tile_index];
-	tileinfo.set(2, code, 0, 0);
+	u16 code = m_reel_vram[layer_num][tile_index*2+1];
+	// colscroll is on upper 8 bits of this (handled in update)
+	u16 color = m_reel_vram[layer_num][tile_index*2];
+	// TODO: confirm if bit 6 is really connected here
+	// upper palette bank is initialized with black at POST and never ever touched again,
+	// not enough to pinpoint one way or another ...
+	tileinfo.set(0, code, color & 0x3f, 0);
 }
 
-void jungleyo_state::bg_videoram_w(offs_t offset, uint16_t data)
+void jungleyo_state::bg_videoram_w(offs_t offset, u16 data, u16 mem_mask)
 {
-	m_bg_videoram[offset / 2] = data;
+	COMBINE_DATA(&m_bg_videoram[offset]);
 	m_bg_tilemap->mark_tile_dirty(offset / 2);
 }
 
-void jungleyo_state::fg_videoram_w(offs_t offset, uint16_t data)
+void jungleyo_state::fg_videoram_w(offs_t offset, u16 data, u16 mem_mask)
 {
-	m_fg_videoram[offset / 2] = data;
+	COMBINE_DATA(&m_fg_videoram[offset]);
 	m_fg_tilemap->mark_tile_dirty(offset / 2);
 }
 
+template <int layer_num> void jungleyo_state::reel_vram_w(offs_t offset, u16 data, u16 mem_mask)
+{
+	COMBINE_DATA(&m_reel_vram[layer_num][offset]);
+	m_reel_tilemap[layer_num]->mark_tile_dirty(offset / 2);
+}
+
 void jungleyo_state::video_start()
 {
 	m_bg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_bg_tile_info)), TILEMAP_SCAN_ROWS, 8, 8, 64, 32);
 	m_fg_tilemap = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_fg_tile_info)), TILEMAP_SCAN_ROWS, 8, 8, 64, 32);
+	m_reel_tilemap[0] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_reel_tile_info<0>)), TILEMAP_SCAN_ROWS, 8, 32, 64, 8);
+	m_reel_tilemap[1] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_reel_tile_info<1>)), TILEMAP_SCAN_ROWS, 8, 32, 64, 8);
+	m_reel_tilemap[2] = &machine().tilemap().create(*m_gfxdecode, tilemap_get_info_delegate(*this, FUNC(jungleyo_state::get_reel_tile_info<2>)), TILEMAP_SCAN_ROWS, 8, 32, 64, 8);
+
 	m_fg_tilemap->set_transparent_pen(0);
+	m_bg_tilemap->set_transparent_pen(0);
+	for (int i = 0;i < 3; i++)
```

Comment:
Missing space after the first semicolon here.  (You could do this with a range-based `for` loop, too – `for (auto &reel : m_reel_tilemap) { reel->… }`).

Reply:
For some reason I forgot that C++ has a _for in_ equivalent, fixed.
I haven't tried any of [these](https://stackoverflow.com/questions/841075/best-c-code-formatter-beautifier) suggestions, any good one that can be also a good candidate for _srcclean_ phase?